### PR TITLE
feat: add `.inspect()` and `.inspect_err()` methods

### PR DESCRIPTION
**Pull request by deliro**
Hello! In Rust 1.76 `inspect` and `inspect_err` methods were stabilized what are very handy when an error (or a value) must be logged. So I implemented them here.

Instead of doing this:

```python
def get_x_from_somewhere() -> Result[..., str]: ...


x = get_x_from_somewhere()
match x:
    case Err(e):
        logger.error("hello world %s", e)

y = x.and_then(upload_to_s3_bucket)
```

You can now do that:

```python
def get_x_from_somewhere() -> Result[..., str]: ...


y = get_x_from_somewhere().inspect_err(lambda e: logger.error("hello world %s", e)).and_then(upload_to_s3_bucket)
```

### CODE REVIEW

**Comment by LEVLLN** [comment]
+1

**Comment by francium** [comment]
Good suggestion. Will review later today

File: tests/test_result.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 from __future__ import annotations
 
-from typing import Callable
+from typing import Callable, List
```
Comment (francium) [inline tests/test_result.py:3]: With the `from __future__ import annotations` already at line 1, you should be able to use `list[...]` directly and no need to import `List`

File: src/result/result.py [hunk 2]
```diff
@@ -202,6 +202,19 @@ def or_else(self, op: object) -> Ok[T]:
         """
         return self
 
+    def inspect(self, op: Callable[[T], None]) -> Result[T, E]:
```
Comment (francium) [inline src/result/result.py:205]: Change all of the `op` types from returning `None` to `Any`, because we shouldn't unnecessarily restrict the user from using a functions that can also return values, we can just ignore any returned value.

```suggestion
    def inspect(self, op: Callable[[T], Any]) -> Result[T, E]:
```

File: tests/test_result.py [hunk 3]
```diff
@@ -213,6 +213,26 @@ def test_and_then() -> None:
     assert Err(3).and_then(sq_lambda).and_then(sq_lambda).err() == 3
 
 
+def test_inspect() -> None:
+    oks: List[int] = []
+    assert Ok(2).inspect(lambda x: oks.append(x)) == Ok(2)
```
Comment (francium) [inline tests/test_result.py:218]: Can you add lines to test using a regular function as well, you'll discover the issue of `None` vs `Any` as I mentioned in my earlier comment,

```python
# Move to bottom of this file along side the other similar functions
def concat_str(x: str) -> str: 
    return x + x

def test_inspect() -> None:
    Err("e").inspect_err(concat_str)
    Err("e").inspect(concat_str)
```

Comment (francium) [inline tests/test_result.py:218]: Can you define these lambda's similar to the `sq_lambda` that is defined at the bottom of this file. Just because it's not possible to annotate the `x` here nicely and my editor gives me poor type checking here currently

image

**Review by francium — COMMENTED** [review]
Thanks for the PR. Just a few small changes are needed and we can get this merged.

File: CHANGELOG.md [hunk 4]
```diff
@@ -13,6 +13,8 @@ Possible log types:
 
 ## [Unreleased]
 
+- `[added]` Add `inspect()` and `inspect_err()` methods
```
[the remark this answers is not in the document]
Comment (deliro) [reply CHANGELOG.md]: done!

File: tests/test_result.py (hunk 1 above)
Comment (deliro) [reply tests/test_result.py]: I wasn't sure if 3.8 has already added this :) 
Done

File: src/result/result.py (hunk 2 above)
Comment (deliro) [reply src/result/result.py]: Yeah, nice take!

File: tests/test_result.py (hunk 3 above)
Comment (deliro) [reply tests/test_result.py]: added a test with a regular function

Comment (deliro) [reply tests/test_result.py]: It can't be moved to the bottom of the file since every fn is a closure there. We can't test a pure function, because a pure function in `inspect*` methods is no-op.
But I type-hinted them

Comment (francium) [reply tests/test_result.py]: You're right. My mistake

**Review by francium — APPROVED** [review]
Thanks for the PR. Looks great.

**Comment by francium** [comment]
Version 0.17.0 published.
